dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs

```diff
@@ -287,16 +277,20 @@ internal KernelFunction(string name, string? pluginName, string description, IRe
                 // Quick check for cancellation after logging about function start but before doing any real work.
                 cancellationToken.ThrowIfCancellationRequested();
 
-                // Invoke pre-invocation filter. If it requests cancellation, throw.
-                var invokingContext = kernel.OnFunctionInvokingFilter(this, arguments);
-
-                if (invokingContext?.Cancel is true)
+                var invocationContext = await kernel.OnFunctionInvocationAsync(this, arguments, (context) =>
                 {
-                    throw new OperationCanceledException("A function filter requested cancellation before function invocation.");
-                }
+                    // Invoke the function and get its streaming enumerable.
+                    var enumerable = this.InvokeStreamingCoreAsync<TResult>(kernel, context.Arguments, cancellationToken);
```

Comment:
When triggering the filter from an AsyncEnumerable we should have a flag on the context so the caller can know when to handle it properly, since we are using the same method.

Reply:
> so the caller can know when to handle it properly instead of need to check Value is a IAsyncEnumerable impl.

I'm not seeing the concern. What is the code someone would write with or without such an additional property? Are you concerned about a non-streaming function returning an IAsyncEnumerable as its result, and the filter confusing that with streaming? Does the distinction matter at that point?

Reply:
@stephentoub We just discussed this internally. So, I believe there are 2 different cases:
- You can invoke a function with `InvokeAsync` and `InvokeStreamingAsync`. `InvokeAsync` will return some result `T`, while `InvokeStreamingAsync` will return `IAsyncEnumerable<T>`.
- `KernelFunctionFromMethod` can return `IAsyncEnumerable<T>` itself.

When I write my filter, I don't know in what way my function was invoked, whether it's `InvokeAsync` or `InvokeStreamingAsync`, which means that in filter every time I will have to check whether I need to override result by simply setting a new value or by using `await foreach` construction. Since same function can be called in both streaming and non-streaming way, the filter will always need to have this check in place.

We were thinking about two different approaches:
- Define a separate type of filter for streaming scenarios (e.g. `IStreamingFunctionFilter`). 
- Define an attribute which you can apply to filter to specify whether its purpose is to work with streaming or non-streaming. No attribute will mean that it will be executed in both cases. Later, this attribute can be extended to specify list of function/plugin names the filter should be triggered for. The attribute should help to avoid `if` conditions to verify whether it's safe to override `IAsyncEnumerable` etc.

What do you think about it?